### PR TITLE
everytrace: New package

### PR DESCRIPTION
**Pull request by citibeth**
Everytrace ensures that stack trace is obtained every time a program exits, for whatever reason.

### CODE REVIEW

**Comment by davydden** [comment]
i must admit I am not in favor of extra `packagename-test` packages. 

@tgamblin @alalazo @adamjstewart : opinions?

File: var/spack/repos/builtin/packages/everytrace/package.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# Install with:
```
Comment (davydden) [inline var/spack/repos/builtin/packages/everytrace/package.py]: copyright header missing. Also this note should probably be removed.

**Comment by alalazo** [comment]
@davydden Maybe tests can be pulled in using the `resource` directive, like `llvm` does for non-core packages ?

**Comment by citibeth** [comment]
If the upstream author put them out as two packages, it's probably best for
Spack to reflect that.

Everytrace is a utility that helps you get stack traces when your program
exits.  Using it requires some setup.  In addition to serving as a test,
`everytrace-test` is a fully worked-out example of how you can use
Everytrace in a CMake-based project.  Had the tests been put into the same
project, then the build would have been different, and it would not be able
to serve as an end-user example.

**Comment by davydden** [comment]
@citibeth :  what about @alalazo 's suggestion above? I agree with him on this one.

**Comment by citibeth** [comment]
I think it's a reasonable suggestion, but probably not the right thing
here.  Imagine you've read about Everytrace and you want to get started
with it in your own project.  As things are now, you can do the following:
```
git clone everytrace-test
cd everytrace-test
spack setup everytrace-test@local
mkdir build
cd build
../spconfig.py ..
make
tests/run_tests   # (or whatever the command is to run the binaries)
```
Once you've worked through that example, you now have a fully-working
Everytrace-based project that you can tweak, extend, copy stuff out of,
etc.  If there's no Spack package for `everytrace-test`, then already
you've made it harder for users to get started with trying out Everytrace,
and seeing how it can work in THEIR projects.

Would it be better if this were called `everytrace-example`?


Spack resources would probably be useful for GALAHAD (see
https://github.com/LLNL/spack/pull/1546 ).  In that case, I have packages
whose sole purpose is to download a tarball and "install" it in a location
where other Spack packages can find it.  Am I correct that Spack resources
address this issue?

The other problem with Spack resources is, as far as I can tell, they are
not yet documented.  Search for `resource` in:
    http://software.llnl.gov/spack/index.html

If someone documents Spack's resource mechanism, I would love to use it for
the GALAHAD stuff.

**Comment by davydden** [comment]
@citibeth i know nothing about `Everytrace`, it was just a general comment. I agree, documentation would be good.

**Comment by citibeth** [comment]
Given the explanation above... do you think I should rename
`everytrace-test` to `everytrace-example`?

**Comment by davydden** [comment]
> Given the explanation above... do you think I should rename `everytrace-test` to `everytrace-example`?

i would say yes.

**Comment by tgamblin** [comment]
@citibeth: I kind of agree with others that it would be nice to have the tests in the same package as the tool.  You're the author of this package, right?  Any chance `ettest` could get pulled in as a build option for `everytrace`?

**Comment by citibeth** [comment]
@tgamblin: The plan is to rename to everytrace-example and leave as a
separate project.  Because it's really an example / template for users
interested in Everytrace, not a unit test.

**Comment by tgamblin** [comment]
@citibeth: I still think examples ought to get bundled with their respective packages... I mean, I guess we're not exactly curating spack packages and there are a lot of them, but I'd rather not have every package show up with a whole suite of small dependent programs.  It seems like this is going to make `spack list` even less manageable... I worry about devolving to `npm`-level every-function-is-a-package insanity.

**Comment by citibeth** [comment]
I felt an example was needed because Everytrace is atypical as a library.
In addition to a `.so` file, Everytrace produces four `.o` files; and you
need to link one of them into YOUR library, depending on which languages
you use and whether use use MPI.  An example that ties this all together in
a CMake context is really helpful.

**Comment by tgamblin** [comment]
@citibeth: I agree i was just suggesting that it be merged into the `everytrace` package as a build option (or just something that comes with it)

**Comment by citibeth** [comment]
@tgamblin:

I'm coming from the "Spack everywhere" point of view, where we never call
CMake manually any more, but instead use `spack setup` to find all our
dependencies for anything we're building.  In most cases, writing a Spack
package is the easiest way to set up your project with the dependencies it
needs.  It should be the first thing a developer does when building a new
piece of software, not the last.  I want developers to see Spack as
something they can work with closely and contribute to as needed, not a
far-off meta-packager to which someone else adds their package if they find
it useful.

If we want to encourage that kind of workflow, we will need examples in
Spack for developers to copy.  `everytrace-example` is just one case; we
probably want more as well.  I'm sure that more will come out of the
tutorial.

In other words... I'm arguing for more example packages in Spack, not
fewer.  We can manage them with tags that we've all agreed are a good
idea.  Just tag examples with the `example` tag and then set `spack find`
and friends to not show them.

Another thing we could do is build in smarts to know that certain packages
are related.  `spack find` and friends would know that `everytrace` and
`everytrace-example` are related, and could show them as one item.  This is
like how UFRaw does things... when you convert a RAW image to .JPG, it
shows the RAW and JPG as one item in its click-list.

Our case is also not so different from that in RHEL, where just about every
`.rpm` has a corresponding `.srpm` for its source code.

So... I think we should merge what we have here as `everytrace-example` and
build out more example features later.

**Comment by citibeth** [comment]
We could do that while preserving the CMake stuff needed in your
Everytrace-enabled project.  But we won't be able to preserve the Spack
package that's also needed (because by definition we'd be eliminating the
example Spack package).

**Comment by tgamblin** [comment]
@citibeth: Ok you've convinced me that slightly finer-grained packages aren't really the worst thing ever.  I have some motivation to make Spack work as a better meta-build tool in addition to being a nice package manager, and finer-grained packages probably have some role to play there.  We can fix this on the UI side if needed.

I'm still convinced that spack should be PEP8-clean, though.  Can you fix the flake errors and then I'll merge?

**Comment by davydden** [comment]
> The plan is to rename to everytrace-example and leave as a
separate project.  Because it's really an example / template for users
interested in Everytrace

one could still install it within `everytrace` into `share/everytrace/examples` or alike. 

I would argue if more examples are needed, they should be added as a resource for a given package and possible associate some variants, like `+examples-type1`, `+examples-type2`, if there is a need to let users choose which examples to install along the package. Otherwise one could always install all examples, which can't hurt. 

>  I still think examples ought to get bundled with their respective packages... I mean, I guess we're not exactly curating spack packages and there are a lot of them, but I'd rather not have every package show up with a whole suite of small dependent programs.

:+1:

@citibeth : there is nothing wrong with what you suggest, I am just of opinion that things should be grouped into a single `package.py` if possible. My understanding is that it is in general possible to do here, although i understand that you disagree on this point.

**Comment by davydden** [comment]
since @tgamblin is convinced and I don't plan to use `everytrace`, I am certainly don't object merging this :smile:

**Comment by tgamblin** [comment]
@davydden: I can see it both ways.  Variants are the more traditional way.  One thing I noticed about the way @citibeth has grouped things is that she's actually using `depends_on` for the example, which tells Spack that if you build the example, you need to build the fortran bindings.  Currently there isn't a great way to do that _within_ a package (variants don't have dependency relationships b/c we don't currently model that).  At least from that angle, using a package actually seems more natural.

Given that we'll likely have thousands of packages, I'm not sure I should complain about clutter in the list.  People are going to query stuff, and a few more packages in a small query result isn't a big deal (and might be informative).  MacPorts for example has over 10k packages... I've never really been concerned one way or the other about the readability of the full list b/c I never actually look at it.

Anyway, that's how I convinced myself.

**Comment by davydden** [comment]
fair points :+1:

File: var/spack/repos/builtin/packages/everytrace-example/package.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+class EverytraceExample(CMakePackage):
+    """Get stack trace EVERY time a program exits."""
+
+    homepage = "https://github.com/citibeth/everytrace-example"
+    version('develop',
+            git='https://github.com/citibeth/everytrace-example.git',
+            branch='develop')
+
+    depends_on('cmake')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/everytrace-example/package.py]: `type='build'`?

**Comment by citibeth** [comment]
OK... after all that ink, the conclusion was to renamed everytrace-test to everytrace-example.  Which I did.  Should be ready to merge now.

One final thing to point out, that we all forgot in the heat of the moment: Everytrace is "special" because the Spack contributor is the same as the upstream author.  In most cases so far, that has not been the case.  If we don't like how the upstream software was packaged, we have no guarantee of being able to get the upstream author to change things more to our liking.  Therefore, Spack in general needs to be flexible with upstream stuff (as we've all discovered, given the zillion weird builds we've encountered).

**Pull request merged**